src/auth_passwd.c

```diff
@@ -65,13 +65,14 @@ static int module_init(Display *display) {
 }
 
 static int module_authenticate(const char *pass) {
+    char *cpw;
 
     if (pass == NULL || pwd_entry == NULL)
         return -1;
 
     /* Simpler, and should work with crypt() algorithms using longer
      * salt strings (like the md5-based one on freebsd).  --marekm */
-    return strcmp(crypt(pass, pwd_entry->pw_passwd), pwd_entry->pw_passwd);
+    return !(cpw = crypt(pass, pwd_entry->pw_passwd)) || strcmp(pwd_entry->pw_passwd, (const char *)cpw) != 0;
```

Comment:
Please, write check for crypt as a separate `if` statement. Also, I think that cast for `const char *` is not required here.

Reply:
Rewritten.